### DISCUSSION TITLE
ERROR foundry_compilers::resolver: failed to resolve versions

### DISCUSSION THREAD

**Discussion by deepakA18** [Cyfrin/foundry-full-course-cu#1816] [Q&A]
I am facing a difficulty due to usage of different libraries(chainlink-brownie-contracts, forge-std and openzeppelin) with different versions how to solve the below error which says ''ERROR foundry_compilers::resolver: failed to resolve versions''

The error looks like:

'''
Error: 
Found incompatible Solidity versions:
script/DeployDSC.s.sol ^0.8.19 imports:
    lib/forge-std/src/Script.sol >=0.6.2, <0.9.0
    src/DecentralisedStableCoin.sol ^0.8.19
    src/DSCEngine.sol ^0.8.19
    script/HelperConfig.s.sol ^0.8.19
    lib/forge-std/src/console.sol >=0.4.22, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/safeconsole.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdChains.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdCheats.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdJson.sol >=0.6.0, <0.9.0
    lib/forge-std/src/StdMath.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStyle.sol >=0.4.22, <0.9.0
    lib/forge-std/src/StdUtils.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Base.sol >=0.6.2, <0.9.0
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20Burnable.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/access/Ownable.sol ^0.8.20
    src/DecentralisedStableCoin.sol ^0.8.19
    lib/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol ^0.8.0
    lib/forge-std/src/Script.sol >=0.6.2, <0.9.0
    test/mocks/MockV3Aggregator.sol ^0.8.19
    lib/openzeppelin-contracts/contracts/mocks/token/ERC20Mock.sol ^0.8.20
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/interfaces/IMulticall3.sol >=0.6.2, <0.9.0
    lib/forge-std/src/mocks/MockERC20.sol >=0.6.2, <0.9.0
    lib/forge-std/src/mocks/MockERC721.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV2V3Interface.sol ^0.6.0
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol ^0.8.20
    lib/forge-std/src/interfaces/IERC20.sol >=0.6.2
    lib/forge-std/src/interfaces/IERC721.sol >=0.6.2
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/interfaces/draft-IERC6093.sol ^0.8.20
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorInterface.sol ^0.6.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol ^0.6.0
    lib/forge-std/src/interfaces/IERC165.sol >=0.6.2
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
Found incompatible Solidity versions:
script/HelperConfig.s.sol ^0.8.19 imports:
    lib/forge-std/src/Script.sol >=0.6.2, <0.9.0
    test/mocks/MockV3Aggregator.sol ^0.8.19
    lib/openzeppelin-contracts/contracts/mocks/token/ERC20Mock.sol ^0.8.20
    lib/forge-std/src/console.sol >=0.4.22, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/safeconsole.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdChains.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdCheats.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdJson.sol >=0.6.0, <0.9.0
    lib/forge-std/src/StdMath.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStyle.sol >=0.4.22, <0.9.0
    lib/forge-std/src/StdUtils.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Base.sol >=0.6.2, <0.9.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV2V3Interface.sol ^0.6.0
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol ^0.8.20
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/interfaces/IMulticall3.sol >=0.6.2, <0.9.0
    lib/forge-std/src/mocks/MockERC20.sol >=0.6.2, <0.9.0
    lib/forge-std/src/mocks/MockERC721.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorInterface.sol ^0.6.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol ^0.6.0
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/interfaces/draft-IERC6093.sol ^0.8.20
    lib/forge-std/src/interfaces/IERC20.sol >=0.6.2
    lib/forge-std/src/interfaces/IERC721.sol >=0.6.2
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    lib/forge-std/src/interfaces/IERC165.sol >=0.6.2
Found incompatible Solidity versions:
test/mocks/MockV3Aggregator.sol ^0.8.19 imports:
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV2V3Interface.sol ^0.6.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorInterface.sol ^0.6.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol ^0.6.0
Found incompatible Solidity versions:
test/unit/DSCEngineTest.t.sol ^0.8.19 imports:
    lib/forge-std/src/Test.sol >=0.6.2, <0.9.0
    script/DeployDSC.s.sol ^0.8.19
    src/DecentralisedStableCoin.sol ^0.8.19
    src/DSCEngine.sol ^0.8.19
    script/HelperConfig.s.sol ^0.8.19
    lib/forge-std/src/console.sol >=0.4.22, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/safeconsole.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdAssertions.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdChains.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdCheats.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdError.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdInvariant.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdJson.sol >=0.6.0, <0.9.0
    lib/forge-std/src/StdMath.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStyle.sol >=0.4.22, <0.9.0
    lib/forge-std/src/StdToml.sol >=0.6.0, <0.9.0
    lib/forge-std/src/StdUtils.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Base.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Script.sol >=0.6.2, <0.9.0
    src/DecentralisedStableCoin.sol ^0.8.19
    src/DSCEngine.sol ^0.8.19
    script/HelperConfig.s.sol ^0.8.19
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20Burnable.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/access/Ownable.sol ^0.8.20
    src/DecentralisedStableCoin.sol ^0.8.19
    lib/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol ^0.8.0
    lib/forge-std/src/Script.sol >=0.6.2, <0.9.0
    test/mocks/MockV3Aggregator.sol ^0.8.19
    lib/openzeppelin-contracts/contracts/mocks/token/ERC20Mock.sol ^0.8.20
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/interfaces/IMulticall3.sol >=0.6.2, <0.9.0
    lib/forge-std/src/mocks/MockERC20.sol >=0.6.2, <0.9.0
    lib/forge-std/src/mocks/MockERC721.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/console.sol >=0.4.22, <0.9.0
    lib/forge-std/src/console2.sol >=0.4.22, <0.9.0
    lib/forge-std/src/safeconsole.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdChains.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdCheats.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdJson.sol >=0.6.0, <0.9.0
    lib/forge-std/src/StdMath.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStorage.sol >=0.6.2, <0.9.0
    lib/forge-std/src/StdStyle.sol >=0.4.22, <0.9.0
    lib/forge-std/src/StdUtils.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Vm.sol >=0.6.2, <0.9.0
    lib/forge-std/src/Base.sol >=0.6.2, <0.9.0
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV2V3Interface.sol ^0.6.0
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol ^0.8.20
    lib/forge-std/src/interfaces/IERC20.sol >=0.6.2
    lib/forge-std/src/interfaces/IERC721.sol >=0.6.2
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/utils/Context.sol ^0.8.20
    lib/openzeppelin-contracts/contracts/interfaces/draft-IERC6093.sol ^0.8.20
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorInterface.sol ^0.6.0
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol ^0.6.0
    lib/forge-std/src/interfaces/IERC165.sol >=0.6.2
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol ^0.8.20
    
  '''
  
  In addition I'm using version '0.8.25'
  
  
  foundry.toml file:
  
  '''
  
  [profile.default]
solc_version= "0.8.25"
src = "src"
out = "out"
libs = ["lib"]
remappings = [
    "@chainlink/contracts=lib/chainlink-brownie-contracts/contracts/",
    "@openzeppelin/contracts=lib/openzeppelin-contracts/contracts"
]

'''

**Comment by EngrPips** [comment]
Hello @deepakA18, why did you explicitly set `solc_version` in `foundry.toml` as `0.8.25`, but all of your contracts which you wrote by yourself are not explicitly using `0.8.25`

```js
src/DecentralisedStableCoin.sol ^0.8.19
src/DSCEngine.sol ^0.8.19
script/HelperConfig.s.sol ^0.8.19
script/DeployDSC.s.sol ^0.8.19
```
